src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
                 return null;
             }
         }
+
+        public static bool VisitBinaryOperatorInterpolatedString<TArg, TInterpolatedStringType>(
+            this BoundBinaryOperator binary,
+            TArg arg,
+            Func<TInterpolatedStringType, TArg, bool> visitor,
+            Action<BoundBinaryOperator, TArg>? binaryOperatorCallback = null)
+            where TInterpolatedStringType : BoundExpression
+        {
+            Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundUnconvertedInterpolatedString) || typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString));
+            var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+
+            pushLeftNodes(binary, stack);
+
+            while (stack.TryPop(out BoundBinaryOperator? current))
+            {
+                switch (current.Left)
+                {
+                    case BoundBinaryOperator op:
+                        binaryOperatorCallback?.Invoke(op, arg);
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!visitor(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Left.Kind);
+                }
+
+                switch (current.Right)
+                {
+                    case BoundBinaryOperator rightOperator:
+                        binaryOperatorCallback?.Invoke(rightOperator, arg);
```

Comment:
>binaryOperatorCallback?.Invoke(rightOperator, arg); [](http://example.com/codeflow?start=24&length=51)

It looks like the algorith is not symmetrical. For some binary operators it invokes callback after invoking visitor  and the callbacks for children, but for others it invokes the callback before.